### DISCUSSION TITLE
how to enable token based  read query  using curl,

### DISCUSSION THREAD

**Discussion by sci110680** [PostgREST/postgrest#3615] [Q&A] [answered]
Hi Team,

I noticed that the PostgREST API currently allows read queries without any authentication or token. For example, using the curl command, data can be fetched directly:

curl http://localhost:3000/todos

Is there a way to enforce authentication so that data cannot be read without a token? If so, could you please provide the necessary steps to configure this?

reference: https://postgrest.org/en/stable/tutorials/tut0.html

Regards,
Ajay

**Comment by wolfgangwalther** [comment, marked as answer]
Have a look at the `db-anon-role` config setting: https://postgrest.org/en/stable/references/configuration.html#db-anon-role

> When unset anonymous access will be blocked.

**Reply by sci110680** [reply]
below parameter is set-
~~~~~~~
[postgres@illin5568 postgrest_token]$ cat tutorial.conf
db-uri = "postgres://authenticator:authenticator@illin5568:5432/paaspg"
db-schema = "api"
db-anon-role = "todo_user"
server-host = "10.234.52.147"
server-port = 3000
jwt-secret = "SY0jCT9KHUAKPVzFjzeEMtMQx6HMCaZU"
~~~~~~~~~~~
still  output is  coming and not asking token .

[postgres@illin5568 postgrest_token]$ **curl "http://illin5568:3000/todos"**
[{"id":1,"done":false,"task":"finish tutorial 0","due":null},
 {"id":2,"done":false,"task":"pat self on back","due":null}]

**Reply by wolfgangwalther** [reply]
Did you actually read what I wrote above?

> When **unset** anonymous access will be blocked.

Of course anonymous access works when you have it set.. that's the whole point of the config.

You need to **unset** it to block anonymous access.

**Reply by sci110680** [reply]
When i  am  unsetting it is  giving below error 
[postgres@illin5568 postgrest_token]$  ./postgrest tutorial.conf
ConfigError {configErrorLocation = KeyMissing ["db-anon-role"], configConversionError = Just [ConversionError {conversionErrorLoc = "requiredValue", conversionErrorWhy = MissingValue, conversionErrorVal = Nothing, conversionErrorType = Just Text, conversionErrorMsg = Nothing}]}

**Reply by wolfgangwalther** [reply]
What does `./postgrest --version` tell you?

**Reply by sci110680** [reply]
postgres@illin5568 postgrest_token]$ ./postgrest --version
Invalid option `--version'

Usage: postgrest FILENAME
  PostgREST 5.1.0 (UNKNOWN) / create a REST API to an existing Postgres database

Available options:
  -h,--help                Show this help text
  FILENAME

**Reply by wolfgangwalther** [reply]
Your PostgREST version was released in September 2018, so is almost 6 years old. You need to update PostgREST...

**Reply by sci110680** [reply]
Thanks  a  lot  , after using latest postgrest-v12.2.0-linux-static-x64.tar.xz .  it is  working .

Now  i get correct message -

without token:
[postgres@illin5568 postgrest_token]$ curl http://illin5568:3000/todos
{"code":"PGRST302","details":null,"hint":null,"message":"Anonymous access is disabled"}

With Token:
[postgres@illin5568 postgrest_token]$ curl http://illin5568:3000/todos -H "Authorization: Bearer $TOKEN"
[{"id":1,"done":false,"task":"finish tutorial 0","due":null},
 {"id":2,"done":false,"task":"pat self on back","due":null}][postgres@illin5568 postgrest_token]$



Thanks a   lot for help .

db-uri = "postgres://authenticator:authenticator@illin5568:5432/paaspg"
db-schema = "api"
**#db-anon-role = "todo_user"**
server-host = "10.234.52.147"
#server-host = "illin5568"
server-port = 3000
jwt-secret = "SY0jCT9KHUAKPVzFjzeEMtMQx6HMCaZU"
